Interface base class

alpaca-tc opened the pull request:
Interfaces should be inherit from BaseClass because those have same methods.
Currently, all interfaces inherits from `LanguageServer::Protocol::Interface::Base`. it have common methods.

- `#attr_requied_keys` defines required keys. If initializing instance without required keys, it raises `ArgumentError`.
- `#attr_optional_keys` defines optional keys.
- Those methods defines accessors. 
   - `attr_optional_keys(:id)` => `instance.id`
- Those methods convert key to lower camel case.
  - `instance.new(underscore_key: '1')` => `instance.attributes.key?(:underscoreKey) == true`
- Some interfaces inherit from parent interface

alpaca-tc commented on lib/language_server/protocol/interface/base.rb:
```diff
@@ -0,0 +1,68 @@
+          lower_camel_case = Protocol::Utils.to_lower_camerize(key.to_s)
+
+          class_eval(<<-METHOD, __FILE__, __LINE__ + 1)
+            def #{key}; @attributes[:#{lower_camel_case}] end
```
Use `#[]` instead of `#fetch` because instance's attributes are already validated.

alpaca-tc commented on lib/language_server/protocol/interface/completion_item.rb:
```diff
@@ -1,144 +1,10 @@
-          attributes.to_json(*args)
-        end
+      class CompletionItem < Base
+        attr_required_keys:label
```
ooops

mtsmfm replied:
Ah, you modified these files manually?

I generated the files by this script:
https://github.com/mtsmfm/language_server-protocol-ruby/blob/fcff46db159de0bba8214ab82abc93b6c940e3e9/scripts/generateFiles.ts

The entrypoint is here:

https://github.com/mtsmfm/language_server-protocol-ruby/blob/fcff46db159de0bba8214ab82abc93b6c940e3e9/bin/generate_files

mtsmfm commented:
Thank you for the great PR!
I love this approach because it's obviously readable ❤️ 

However, I just remembered why I chose this verbose approach.
(Sorry being late to tell you 🙇 )
I think (and hope) our language server can extract/understand keyword arguments and the document from these files.

So I want to leave documents at least and hopefully want to use keyword args.

alpaca-tc commented:
Okay i can try it 😄 👍

alpaca-tc commented:
@mtsmfm I fixed the keyword arguments 😄 
BTW I just notified that some interfaces inherit from parent interface; you can find them [here](https://github.com/Microsoft/language-server-protocol/blob/master/protocol.md#textDocument_hover) `open search area <C-F> and search 'extends'` . In addition I fixed it. 👍

mtsmfm commented:
> BTW I just notified that some interfaces inherit from parent interface;

Oops 😵 
Thanks!

mtsmfm commented:
How did you modify these files?

https://github.com/mtsmfm/language_server-protocol-ruby/pull/4#discussion_r140802760

mtsmfm commented:
> So I want to leave documents at least and hopefully want to use keyword args.

I meant "we should *keep* document for attributes"
For example, `The request id`, `The method to be invoked.` and `The method's params.` should remain I think.

```typescript
interface RequestMessage extends Message {

	/**
	 * The request id.
	 */
	id: number | string;

	/**
	 * The method to be invoked.
	 */
	method: string;

	/**
	 * The method's params.
	 */
	params?: any
}
```

alpaca-tc commented on scripts/generateFiles.ts:
```diff
@@ -148,33 +148,19 @@ module LanguageServer
-
-          @attributes.freeze
-        end
+      class {{definition.interface.name}} < Base
```
Unfortunately, it is difficult for me to inspect parent class. 👀 💧
